### PR TITLE
Polyester for threading?

### PR DESCRIPTION
**Pull request by Moelf**
https://github.com/JuliaSIMD/Polyester.jl/issues/44 thanks @chriselrod

for workload similar to https://github.com/Moelf/UnROOT_RDataFrame_MiniBenchmark , this speeds up to 2x (because we're IO bounded), it should do much better if computation is actually intensive. The good thing about Polyester.jl is that we can customize the behavior in complicated way.

To use it, simply:
```julia
@batch for evt in mytree
	...
end
```

```
# for the MiniBenchmark
- plain: 23.67 seconds
- @sync @async: 24.08 seconds
- @batch: 7.607535 seconds
```

### CODE REVIEW

**Comment by Moelf** [comment]
we should includes this optionally, to not break on lower version

**Comment by chriselrod** [comment]
As Polyester 0.4 no longer depends (directly or indirectly) on VectorizationBase, it probably wouldn't be too difficult to add Julia 1.3 support.

**Comment by Moelf** [comment]
```julia
Type stability |    2      2
Test Summary:                    | Pass  Total
Parallel and enumerate interface |   17     17
     Testing UnROOT tests passed 

julia> Threads.nthreads()
3
```

can't reproduce error locally, will investigate

**Comment by aminnj** [comment]
Based on [this](https://discourse.julialang.org/t/correct-use-of-requires-jl/10493/3), the user still has to type `using Polyester` in order to use `@batch` because of Requires.jl? :(

**Comment by aminnj** [comment]
I can't reproduce it locally either. It's even more odd because that test file only has 1 basket (events 0 to 1000) so we shouldn't have boundary issues or pull from the wrong thread local cache (they're all identical).

**Comment by aminnj** [comment]
OK I think I figured it out the solution but I don't know the problem still. Comment out these three lines and the tests should succeed:
https://github.com/tamasgal/UnROOT.jl/blob/42ddc0b13067c5c96bb2d7bf9853920c73a1bbec/test/runtests.jl#L552-L554

You can reproduce it locally actually. Store the list of event indices each thread worked on. In theory they should be disjoint. But we find as we keep doing these loops, the sets start getting common elements. It doesn't happen in the REPL since you need the above 3 lines in a test set. Some kind of issue with multithreading a test set?

```julia
function printdisjoint(s)
    for i in 1:length(s)
        for j in 1:length(s)
            j <= i && continue
            common = s[i] ∩ s[j]
            println("thread=$i thread=$j ncommon=$(length(common)): $(repr(common, context=:limit => true))")
        end
    end
end

for i in 1:3
    println("ITER $i")
    indices = [Vector{Int}() for _ in 1:Threads.nthreads()]
    @batch for (i,evt) in enumerate(t)
        push!(indices[Threads.threadid()], i)
    end
    printdisjoint(indices)
end
```
output:
```
ITER 1
thread=1 thread=2 ncommon=0: Int64[]
thread=1 thread=3 ncommon=0: Int64[]
thread=1 thread=4 ncommon=0: Int64[]
thread=2 thread=3 ncommon=0: Int64[]
thread=2 thread=4 ncommon=0: Int64[]
thread=3 thread=4 ncommon=0: Int64[]
ITER 2
thread=1 thread=2 ncommon=4: [798, 7, 957, 997]
thread=1 thread=3 ncommon=66: [252, 257, 821, 261, 823, 825, 265, 266, 829, 270  …  329, 330, 331, 332, 333, 899, 335, 901, 902, 905]
thread=1 thread=4 ncommon=3: [619, 948, 997]
thread=2 thread=3 ncommon=2: [345, 714]
thread=2 thread=4 ncommon=102: [503, 506, 12, 509, 510, 17, 18, 21, 24, 26  …  696, 697, 698, 700, 701, 237, 239, 709, 244, 714]
thread=3 thread=4 ncommon=10: [345, 706, 714, 722, 416, 425, 432, 436, 440, 455]
ITER 3
thread=1 thread=2 ncommon=16: [264, 762, 764, 26, 29, 790, 75, 881, 892, 898, 926, 185, 199, 209, 212, 975]
thread=1 thread=3 ncommon=8: [264, 315, 328, 416, 428, 466, 471, 494]
thread=1 thread=4 ncommon=13: [763, 26, 572, 75, 876, 881, 892, 898, 926, 185, 209, 212, 975]
thread=2 thread=3 ncommon=22: [253, 257, 264, 274, 278, 298, 335, 76, 352, 378  …  410, 422, 426, 435, 442, 443, 446, 450, 468, 489]
thread=2 thread=4 ncommon=122: [1, 6, 507, 13, 18, 19, 20, 21, 274, 278  …  213, 215, 216, 219, 222, 739, 975, 239, 240, 242]
thread=3 thread=4 ncommon=20: [274, 277, 278, 282, 298, 300, 335, 352, 378, 393, 398, 404, 405, 426, 435, 442, 450, 458, 461, 489]
```

**Comment by aminnj** [comment]
Here's a minimal reproducer
```julia
using Test
using UnROOT
const SAMPLES_DIR = joinpath(@__DIR__, "samples")

# comment out the 2 testset wrapper lines to make tests pass
@testset "Enumerate interface" begin
    t = LazyTree(ROOTFile(joinpath(SAMPLES_DIR, "NanoAODv5_sample.root")), "Events", ["Muon_pt"])
    for i in 1:3
        println("ITER $i")
        indices = [Vector{Int}() for _ in 1:Threads.nthreads()]
        @batch for (i,evt) in enumerate(t)
            push!(indices[Threads.threadid()], i)
        end
        @test sum([length(indices[i] ∩ indices[j]) for i=1:length(indices), j=1:length(indices) if j>i]) == 0
    end
    # OR comment out this line to make tests pass
    i, evt = enumerate(t)[2]
end
```

**Comment by Moelf** [comment]
what...that's so weird, time-travel? In all seriousness, one should never directly indexing a `enumerate`

**Comment by Moelf** [comment]
manual review request @aminnj :rofl:

File: test/runtests.jl [hunk 1]
```diff
@@ -543,14 +545,36 @@ end
...
+        end
+        @test count(>(0), event_nums) > 1 # test @batch is actually threading
+        @test sum(event_nums) == length(t)
+
+        nmus .= 0
+        @batch for evt in t
+            nmus[Threads.threadid()] += length(evt.Muon_pt)
+        end
+        @test count(>(0), nmus) > 1 # test @batch is actually threading
         @test sum(nmus) == 878
```
Comment (aminnj) [inline test/runtests.jl:570]: just to make sure we don't have any weirdness in the future, could you add
```julia
for i in 1:3
    inds = [Vector{Int}() for _ in 1:Threads.nthreads()]
    @batch for (i,evt) in enumerate(t)
        push!(inds[Threads.threadid()], i)
    end
    @test sum([length(inds[i] ∩ inds[j]) for i=1:length(inds), j=1:length(inds) if j>i]) == 0
end
```

File: docs/src/performancetips.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+## `Threads.@threads` should go with `enumerate()`
+Unlike `@batch` from polyester, there's not much we can do to customize behavior of `@threads`. It is essentially
+calling `getindex()`, which we want to keep eager for regular use (e.v `mytree[120]` is eager). Thus, if for some
+reason you want to use `@threads` instead of `@batch`, you should use it with `enumerate`:
+```julia
+# check Threads.nthreads() > 1
+Threads.@threads for (_, evt) in enumerate(mytree)
+    ...
+end
+```
```
Comment (aminnj) [inline docs/src/performancetips.md:28]: To clearly show eager vs noneager, you can include
```julia
julia> for evt in mytree
           @show evt
           break
       end
evt = "LazyEvent with: (:tree, :idx)"

julia> Threads.@threads for evt in mytree
           @show evt
           break
       end
evt = (nMuon = 0x00000000, Muon_pt = Float32[])
evt = (nMuon = 0x00000001, Muon_pt = Float32[3.4505641])
evt = (nMuon = 0x00000000, Muon_pt = Float32[])
evt = (nMuon = 0x00000002, Muon_pt = Float32[21.279676, 7.6710315])
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -43,7 +43,7 @@ jobs:
       - uses: julia-actions/julia-buildpkg@v1
       - uses: julia-actions/julia-runtest@v1
         env:
-          JULIA_NUM_THREADS: "3"
+          JULIA_NUM_THREADS: "2"
```
Comment (aminnj) [inline .github/workflows/ci.yml:46]: Can we bump this up again to increase chances of finding race conditions maybe?
